src/libraries/Microsoft.PowerFx.Interpreter/ReflectionFunction.cs

```diff
@@ -207,21 +209,36 @@ private FunctionDescr Scan()
                 info.RetType = GetType(m.ReturnType);
 
                 var paramTypes = new List<FormulaType>();
-                foreach (var p in m.GetParameters())
+
+                info._isAsync = m.ReturnType.BaseType == typeof(Task);
+
+                var parmas = m.GetParameters();
+                for (var i = 0; i < parmas.Length; i++)
                 {
-                    if (typeof(FormulaValue).IsAssignableFrom(p.ParameterType))
+                    if (i == parmas.Length - 1 && info._isAsync)
                     {
-                        paramTypes.Add(GetType(p.ParameterType));
-                    } 
-                    else if (p.ParameterType == ConfigType)
+                        if (parmas[i].ParameterType != typeof(CancellationToken))
+                        {
+                            throw new InvalidOperationException($"Last argument must be a cancellation token.");
+                        }
+                    }
+                    else if (typeof(FormulaValue).IsAssignableFrom(parmas[i].ParameterType))
+                    {
+                        paramTypes.Add(GetType(parmas[i].ParameterType));
+                    }
+                    else if (parmas[i].ParameterType == ConfigType)
                     {
                         // Not a Formulatype, pull from RuntimeConfig
-                        info._configType = p.ParameterType;
-                    } 
-                    else
+                        info._configType = parmas[i].ParameterType;
+                    }
+                    else if (parmas[i].ParameterType == typeof(CancellationToken) && _info._isAsync)
                     {
-                        // Unknonw parameter type
-                        throw new InvalidOperationException($"Unknown parameter type: {p.Name}, {p.ParameterType}");
+                        throw new InvalidOperationException($"Cancellation token must be the last argument.");
```

Comment:
>InvalidOperationException [](http://example.com/codeflow?start=34&length=25)

why do we have this check on both line 236 and line 222?

Reply:
222 checks if the last argument is Cancellation token.
while 236 is in-case user passed cancellation token somewhere in middle (in-case they don't know convention) and we want to give the better feedback in error message.